lib/puppet/util/log.rb

```diff
@@ -244,39 +255,36 @@ def self.from_pson(data)
     self.from_data_hash(data)
   end
 
-  attr_accessor :time, :remote, :file, :line, :source
-  attr_reader :level, :message
+  attr_reader :level, :message, :issue_code, :time, :file, :line, :pos, :backtrace, :source, :node, :environment
 
   def initialize(args)
-    self.level = args[:level]
-    self.message = args[:message]
-    self.source = args[:source] || "Puppet"
+    level = args[:level]
+    raise ArgumentError, "Puppet::Util::Log requires a log level" if level.nil?
+    raise ArgumentError, "Puppet::Util::Log requires that log level is a symbol or string" unless level.respond_to? "to_sym"
 
-    @time = Time.now
+    level = level.to_sym
+    raise ArgumentError, "Invalid log level #{level}" unless self.class.validlevel?(level)
+    @level = level
+    # Tag myself with my log level
+    tag(level)
 
-    if tags = args[:tags]
-      tags.each { |t| self.tag(t) }
-    end
+    message = args[:message]
+    raise ArgumentError, "Puppet::Util::Log requires a message" if message.nil?
+    @message = message.to_s
 
-    [:file, :line].each do |attr|
-      next unless value = args[attr]
-      send(attr.to_s + "=", value)
+    # Avoid setting these instance variables. We don't want them defined unless they exist
+    [:backtrace, :environment, :node, :issue_code, :file, :line, :pos].each do |attr|
+      value = args[attr]
+      instance_variable_set("@#{attr}", value) unless value.nil?
     end
 
-    Log.newmessage(self)
```

Comment:
This removal changes the API/semantics of Log. If there is code that logs by calling Log.new (and not create) the result is silently dropping those entries. The original construct is horrible as it means it is logged immediately when creating the Log instance. I wonder if we need to make this conditional - when coming from create it passes an additional arg :do_not_log - thereby preserving the old API ? (+ log issue to clean this up asap on major release boundary - 4.0 if we can)